File: python/1-50/10. Regular Expression Matching.py
class Solution(object):

    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        def helper(s, p, s_index, p_index):
            if s_index == -1 and p_index == -1:
                return True
            if p_index == -1 and s_index != -1:
                return False
            if p_index != -1 and s_index == -1:
                if p_index >= 2 and p[p_index - 1] == "*":
                    return helper(s, p, s_index, p_index - 2)
                else:
                    return False
            
        return helper(s, p, len(s) - 1, len(p) - 1)

sol = Solution()
print sol.isMatch("aa", "a")

File: cpp/1-50/10. Regular Expression Matching.cpp
class Solution 
{
public:
    bool isMatch(string s, string p) 
    {
        vector<vector<bool> > table(s.length() + 1, vector<bool>(p.length() + 1));
        table[0][0] = true;
        for(int j = 1; j <= p.length(); j++)
        {
                if(p[j-1] == '*')
                        table[0][j] = table[0][j-2];
        }
        for(int i = 1; i <= s.length(); i++)
        {
                for(int j = 1; j <= p.length(); j++)
                {
                        if(s[i-1] == p[j-1] || p[j-1] == '.')
                                table[i][j] = table[i-1][j-1];
                        if(p[j-1] == '*')
                                table[i][j] = table[i][j-2] | (table[i-1][j] & (p[j-2] == s[i-1]| p[j-2] == '.'));
                }
        }
        return table[s.length()][p.length()];
    }
};